src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -2007,6 +2008,20 @@ internal PointerTypeSymbol CreatePointerTypeSymbol(TypeSymbol elementType, Nulla
             throw new NotImplementedException();
         }
 
+        private ImmutableQueue<SourceMethodSymbolWithAttributes> _moduleInitializerMethods = ImmutableQueue<SourceMethodSymbolWithAttributes>.Empty;
```

Comment:
>ImmutableQueue<SourceMethodSymbolWithAttributes> [](start = 16, length = 48)

I think ```ConcurrentSet<MethodSymbol>``` is a better choice. It might be possible to get in a situation when two threads try to add the same method. We shouldn't fail and shouldn't call the method twice

Reply:
Is it possible to get in a situation where two threads produce separate MethodSymbol instances that ConcurrentSet doesn't consider equal?

Does using MethodSymbol hamper getting the file and line information for the future sort needing for deterministic ordering?

Reply:
> Is it possible to get in a situation where two threads produce separate MethodSymbol instances that ConcurrentSet doesn't consider equal?

Not for the same method declaration in source. We will have to test what happens with two parts of partial methods though, I guess we can do that while working on error reporting.

> Does using MethodSymbol rather than SourceMethodSymbol hamper getting the file and line information for the future sorting that is required for deterministic ordering?

I think we will use LexicalOrderSymbolComparer, which works on any Symbol.

---
In reply to: [414876916](https://github.com/dotnet/roslyn/pull/43281#discussion_r414876916) [](ancestors = 414876916)

Reply:
> We will have to test what happens with two parts of partial methods though, I guess we can do that while working on error reporting.

I skipped testing partial methods in the error reporting PR (waiting for this one to merge) because partial methods are private and private methods can't be added to this collection.

Reply:
> I skipped testing partial methods in the error reporting PR (waiting for this one to merge) because partial methods are private and private methods can't be added to this collection. 

It is still worth having a test.
I am also thinking ahead, there is a feature in the works to "liberate" partial methods, i.e. allow them to have any accessibility, etc.  

---
In reply to: [414889856](https://github.com/dotnet/roslyn/pull/43281#discussion_r414889856) [](ancestors = 414889856)